src/notebooks/controllers/vscodeNotebookController.ts

```diff
@@ -443,27 +485,31 @@ export class VSCodeNotebookController implements Disposable, IVSCodeNotebookCont
         return currentExecution;
     }
 
-    private async executeCell(doc: NotebookDocument, cell: NotebookCell) {
+    private async executeCell(notebook: NotebookDocument, cell: NotebookCell) {
         traceInfo(`Execute Cell ${cell.index} ${getDisplayPath(cell.notebook.uri)}`);
         // Start execution now (from the user's point of view)
         let exec = this.createCellExecutionIfNecessary(cell, this.controller);
 
         // Connect to a matching kernel if possible (but user may pick a different one)
         let currentContext: 'start' | 'execution' = 'start';
         let kernel: IKernel | undefined;
-        let controller = this.controller;
+        let controller: NotebookController = this.controller;
+        let oldController: NotebookController = this.controller;
         let kernelStarted = false;
         try {
-            kernel = await this.connectToKernel(doc, new DisplayOptions(false));
-            kernelStarted = true;
+            kernel = await this.connectToKernel({ notebook }, new DisplayOptions(false));
             // If the controller changed, then ensure to create a new cell execution object.
-            if (kernel && kernel.controller.id !== controller.id) {
+            if (kernel && kernel.controller.id !== oldController.id) {
```

Comment:
it seems like this could just compare to `controller.id` and drop the `oldController` variable